File: 0500-0599/553. Optimal Division/Solution.cpp
//Approach-1 (Greedy + String Construction)
//T.C : O(n)
//S.C : O(n)

class Solution {
public:

    string optimalDivision(
        vector<int>& nums
    ) {

        int n = nums.size();

        if (n == 1) {

            return to_string(
                nums[0]
            );
        }

        if (n == 2) {

            return
                to_string(nums[0])
                + "/"
                + to_string(nums[1]);
        }

        string answer =
            to_string(nums[0])
            + "/("
            + to_string(nums[1]);

        for (
            int i = 2;
            i < n;
            i++
        ) {

            answer += "/";

            answer +=
                to_string(nums[i]);
        }

        answer += ")";

        return answer;
    }
};


/*
//Approach-2 (Recursive DFS / Brute Force)
//T.C : Exponential
//S.C : O(n)

class Solution {
public:

    pair<double, string> getMaximum(
        vector<int>& nums,
        int left,
        int right
    ) {

        if (left == right) {

            return {
                nums[left],
                to_string(nums[left])
            };
        }

        pair<double, string> best = {
            -1.0,
            ""
        };

        for (
            int k = left;
            k < right;
            k++
        ) {

            auto leftPart =
                getMaximum(
                    nums,
                    left,
                    k
                );

            auto rightPart =
                getMinimum(
                    nums,
                    k + 1,
                    right
                );

            double value =
                leftPart.first /
                rightPart.first;

            if (value > best.first) {

                string expression =
                    leftPart.second
                    + "/";

                if (k + 1 < right) {

                    expression +=
                        "("
                        + rightPart.second
                        + ")";
                } else {

                    expression +=
                        rightPart.second;
                }

                best = {
                    value,
                    expression
                };
            }
        }

        return best;
    }

    pair<double, string> getMinimum(
        vector<int>& nums,
        int left,
        int right
    ) {

        if (left == right) {

            return {
                nums[left],
                to_string(nums[left])
            };
        }

        pair<double, string> best = {
            1e18,
            ""
        };

        for (
            int k = left;
            k < right;
            k++
        ) {

            auto leftPart =
                getMinimum(
                    nums,
                    left,
                    k
                );

            auto rightPart =
                getMaximum(
                    nums,
                    k + 1,
                    right
                );

            double value =
                leftPart.first /
                rightPart.first;

            if (value < best.first) {

                string expression =
                    leftPart.second
                    + "/";

                if (k + 1 < right) {

                    expression +=
                        "("
                        + rightPart.second
                        + ")";
                } else {

                    expression +=
                        rightPart.second;
                }

                best = {
                    value,
                    expression
                };
            }
        }

        return best;
    }

    string optimalDivision(
        vector<int>& nums
    ) {

        return getMaximum(
            nums,
            0,
            nums.size() - 1
        ).second;
    }
};
*/

File: 0500-0599/553. Optimal Division/Solution.java
//Approach-1 (Greedy + String Construction)
//T.C : O(n)
//S.C : O(n)

class Solution {

  public String optimalDivision(int[] nums) {

    int n = nums.length;

    if (n == 1) {

      return String.valueOf(
          nums[0]);
    }

    if (n == 2) {

      return nums[0]
          + "/"
          + nums[1];
    }

    StringBuilder answer = new StringBuilder();

    answer.append(nums[0]);

    answer.append("/(");

    answer.append(nums[1]);

    for (int i = 2; i < n; i++) {

      answer.append("/");

      answer.append(nums[i]);
    }

    answer.append(")");

    return answer.toString();
  }
}

/*
 * //Approach-2 (Recursive DFS / Brute Force)
 * //T.C : Exponential
 * //S.C : O(n)
 * 
 * class Solution {
 * 
 * private static class Node {
 * 
 * double value;
 * 
 * String expression;
 * 
 * Node(
 * double value,
 * String expression
 * ) {
 * 
 * this.value = value;
 * 
 * this.expression = expression;
 * }
 * }
 * 
 * public String optimalDivision(int[] nums) {
 * 
 * return getMaximum(
 * nums,
 * 0,
 * nums.length - 1
 * ).expression;
 * }
 * 
 * private Node getMaximum(
 * int[] nums,
 * int left,
 * int right
 * ) {
 * 
 * if (left == right) {
 * 
 * return new Node(
 * nums[left],
 * String.valueOf(nums[left])
 * );
 * }
 * 
 * Node best =
 * new Node(
 * -1,
 * ""
 * );
 * 
 * for (
 * int mid = left;
 * mid < right;
 * mid++
 * ) {
 * 
 * Node leftPart =
 * getMaximum(
 * nums,
 * left,
 * mid
 * );
 * 
 * Node rightPart =
 * getMinimum(
 * nums,
 * mid + 1,
 * right
 * );
 * 
 * double value =
 * leftPart.value
 * /
 * rightPart.value;
 * 
 * if (value > best.value) {
 * 
 * String expression =
 * leftPart.expression
 * + "/";
 * 
 * if (mid + 1 < right) {
 * 
 * expression +=
 * "("
 * + rightPart.expression
 * + ")";
 * } else {
 * 
 * expression +=
 * rightPart.expression;
 * }
 * 
 * best =
 * new Node(
 * value,
 * expression
 * );
 * }
 * }
 * 
 * return best;
 * }
 * 
 * private Node getMinimum(
 * int[] nums,
 * int left,
 * int right
 * ) {
 * 
 * if (left == right) {
 * 
 * return new Node(
 * nums[left],
 * String.valueOf(nums[left])
 * );
 * }
 * 
 * Node best =
 * new Node(
 * Double.MAX_VALUE,
 * ""
 * );
 * 
 * for (
 * int mid = left;
 * mid < right;
 * mid++
 * ) {
 * 
 * Node leftPart =
 * getMinimum(
 * nums,
 * left,
 * mid
 * );
 * 
 * Node rightPart =
 * getMaximum(
 * nums,
 * mid + 1,
 * right
 * );
 * 
 * double value =
 * leftPart.value
 * /
 * rightPart.value;
 * 
 * if (value < best.value) {
 * 
 * String expression =
 * leftPart.expression
 * + "/";
 * 
 * if (mid + 1 < right) {
 * 
 * expression +=
 * "("
 * + rightPart.expression
 * + ")";
 * } else {
 * 
 * expression +=
 * rightPart.expression;
 * }
 * 
 * best =
 * new Node(
 * value,
 * expression
 * );
 * }
 * }
 * 
 * return best;
 * }
 * }
 */